File: sdk/javascript/src/types/monitoring.ts
/**
 * Alert severity enum
 */
export enum AlertSeverity {
  INFO = 0,
  WARNING = 1,
  ERROR = 2,
  CRITICAL = 3,
}

/**
 * Alert status enum
 */
export enum AlertStatus {
  ACTIVE = 0,
  ACKNOWLEDGED = 1,
  RESOLVED = 2,
}

/**
 * System metrics
 */
export interface SystemMetrics {
  blockHeight: number;
  blockTime: number;
  transactionsPerSecond: number;
  activeValidators: number;
  totalStaked: string;
  networkLoad: number;
  memoryUsage: number;
  cpuUsage: number;
  diskUsage: number;
  peerCount: number;
  timestamp: Date;
}

/**
 * Alert
 */
export interface Alert {
  id: string;
  type: string;
  severity: AlertSeverity;
  status: AlertStatus;
  message: string;
  details: Record<string, any>;
  source: string;
  triggeredAt: Date;
  acknowledgedAt?: Date;
  resolvedAt?: Date;
  acknowledgedBy?: string;
}

/**
 * Monitoring configuration
 */
export interface MonitoringConfig {
  enabled: boolean;
  metricsInterval: number;
  alertThresholds: {
    cpuUsage: number;
    memoryUsage: number;
    diskUsage: number;
    blockTime: number;
    transactionRate: number;
  };
  alertRecipients: string[];
  retentionPeriod: number;
}

/**
 * Performance metrics
 */
export interface PerformanceMetrics {
  avgBlockTime: number;
  avgTransactionsPerBlock: number;
  avgGasPrice: string;
  networkThroughput: number;
  validatorUptime: number;
  missedBlocks: number;
  timestamp: Date;
}

/**
 * Node health
 */
export interface NodeHealth {
  nodeId: string;
  status: string;
  uptime: number;
  version: string;
  syncStatus: {
    catching_up: boolean;
    latest_block_height: number;
    latest_block_time: Date;
  };
  peers: number;
  validatorInfo?: {
    address: string;
    votingPower: string;
    proposerPriority: number;
  };
}

/**
 * Anomaly detection result
 */
export interface AnomalyDetection {
  detected: boolean;
  anomalyType: string;
  severity: number;
  confidence: number;
  affectedMetrics: string[];
  timestamp: Date;
  details: string;
}

/**
 * Monitoring parameters
 */
export interface MonitoringParams {
  enabled: boolean;
  metricsRetention: number;
  alertingEnabled: boolean;
  anomalyDetectionEnabled: boolean;
  mlModelVersion: string;
  samplingRate: number;
}

File: sdk/python/aura/types/monitoring.py
"""Type definitions for Monitoring module."""

from dataclasses import dataclass
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class AlertSeverity(Enum):
    """Alert severity levels."""

    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"


class AlertStatus(Enum):
    """Alert status."""

    ACTIVE = "active"
    ACKNOWLEDGED = "acknowledged"
    RESOLVED = "resolved"


class HealthStatus(Enum):
    """System health status."""

    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"
    CRITICAL = "critical"


@dataclass
class MonitoringAlert:
    """Monitoring alert."""

    alert_id: str
    title: str
    description: str
    severity: AlertSeverity
    status: AlertStatus
    source: str
    created_at: datetime
    updated_at: datetime
    acknowledged_by: Optional[str] = None
    resolved_at: Optional[datetime] = None
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class MonitoringMetric:
    """Monitoring metric data."""

    metric_name: str
    value: float
    unit: str
    timestamp: datetime
    labels: Optional[Dict[str, str]] = None
    tags: Optional[List[str]] = None


@dataclass
class SystemHealth:
    """System health information."""

    overall_status: HealthStatus
    components: Dict[str, str]
    uptime: int
    last_check: datetime
    active_alerts: int
    cpu_usage: float
    memory_usage: float
    disk_usage: float
    network_latency: float


@dataclass
class AnomalyDetectionParams:
    """Anomaly detection parameters."""

    metric_name: str
    threshold: float
    window_size: int
    sensitivity: float


@dataclass
class AnomalyEvent:
    """Detected anomaly event."""

    event_id: str
    metric_name: str
    detected_value: float
    expected_range: str
    severity: AlertSeverity
    detected_at: datetime
    confidence: float


@dataclass
class PerformanceMetrics:
    """Performance metrics."""

    average_block_time: float
    transactions_per_second: float
    average_gas_price: str
    network_hashrate: float
    active_validators: int
    timestamp: datetime


@dataclass
class AlertRule:
    """Alert rule configuration."""

    rule_id: str
    name: str
    condition: str
    severity: AlertSeverity
    enabled: bool
    notification_channels: List[str]


@dataclass
class MonitoringDashboard:
    """Dashboard metrics summary."""

    total_nodes: int
    active_nodes: int
    total_transactions: int
    failed_transactions: int
    average_response_time: float
    error_rate: float
    timestamp: datetime
